### ISSUE TITLE
warn if rerun-if-changed=PATH is used but PATH is empty

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/cargo#6003]
I recently ran into a weird bug where my project would always rebuild main.rs even though I didn't change any of its files.

Turned out in ````build.rs```` I had `rerun-if-changed=PATH` set to a wrong `PATH`, a `PATH` that never was any file, which caused to trigger rebuilds every time.

Can we print a warning if the `PATH` is empty, or are there cases where this could be intended?

**Comment by remysucre** [comment]
Another reasonable behavior (perhaps on top of the warning) is to rebuild only when the file is removed between builds or is created just before a new build (i.e. the file is "modified" in the sense that it went from existent to non-existent or vice versa).

**Comment by dtolnay** [comment]
I agree with @remysucre that a warning is not necessarily appropriate.

I came across the spurious rebuilding behavior described in this issue in the case of https://github.com/rust-lang/crates.io &mdash; they have a [`cargo:rerun-if-changed=.env`](https://github.com/rust-lang/crates.io/blob/c32fef25c58e53b007dd5795e39fae48800d9bfe/build.rs#L7) but a `.env` file is not checked into the repo, in fact it is [ignored via gitignore](https://github.com/rust-lang/crates.io/blob/c32fef25c58e53b007dd5795e39fae48800d9bfe/.gitignore#L24).

The behavior I would have expected is that Cargo remembers which rerun-if-changed files did not exist in the most recent build, and does not rebuild if they still do not exist. Could we consider the continual rebuilding behavior a bug?

<pre>
<a href="https://github.com/rust-lang/crates.io">crates.io</a>$  <b>cargo check</b>
<i>...</i>
<i>...</i>
    <i>Finished dev [unoptimized + debuginfo] target(s) in 21.91s</i>

<a href="https://github.com/rust-lang/crates.io">crates.io</a>$  <b>cargo check</b>
   <i>Compiling cargo-registry v0.2.2</i>
    <i>Finished dev [unoptimized + debuginfo] target(s) in 3.21s</i>

<a href="https://github.com/rust-lang/crates.io">crates.io</a>$  <b>cargo check</b>
   <i>Compiling cargo-registry v0.2.2</i>
    <i>Finished dev [unoptimized + debuginfo] target(s) in 3.24s</i>

<a href="https://github.com/rust-lang/crates.io">crates.io</a>$  <b>cargo check</b>                       <kbd><b># each of these is a rebuild</b></kbd>
   <i>Compiling cargo-registry v0.2.2</i>
    <i>Finished dev [unoptimized + debuginfo] target(s) in 3.19s</i>

<a href="https://github.com/rust-lang/crates.io">crates.io</a>$  <b>touch .env</b>

<a href="https://github.com/rust-lang/crates.io">crates.io</a>$  <b>cargo check</b>
   <i>Compiling cargo-registry v0.2.2</i>
    <i>Finished dev [unoptimized + debuginfo] target(s) in 3.20s</i>

<a href="https://github.com/rust-lang/crates.io">crates.io</a>$  <b>cargo check</b>
    <i>Finished dev [unoptimized + debuginfo] target(s) in 0.12s</i>

<a href="https://github.com/rust-lang/crates.io">crates.io</a>$  <b>cargo check</b>
    <i>Finished dev [unoptimized + debuginfo] target(s) in 0.12s</i>

<a href="https://github.com/rust-lang/crates.io">crates.io</a>$  <b>cargo check</b>                       <kbd><b># no longer rebuilding now that .env exists</b></kbd>
    <i>Finished dev [unoptimized + debuginfo] target(s) in 0.12s</i>
</pre>

**Comment by RalfJung** [comment]
In https://github.com/rust-lang/miri/pull/4453 we would also have benefited from the "remember that the directory did not exist" version of this, i.e. causing a rebuild once the dir does exist but until then treating continued non-existence as a non-change.

> It won't be trivial to implement because cargo does not track the status of every file (it just tracks the newest mtime).

The ideal representation for the behavior suggested above would be to track `Option<mtime>` so `None` can be recorded as a tracked result. But I guess it could be represented as an mtime of 0 or so?
